File: WINGS_AI_Bootcamp/Elixir/ClassifyStrings.ex
defmodule Classifier do
  def relation(a, b) when a < b, do: "<"
  def relation(a, b) when a == b, do: "="
  def relation(_a, _b), do: ">"

  def squeeze(str) do
    do_squeeze(String.graphemes(str))
    |> Enum.join()
  end

  defp do_squeeze([a, b | rest]) when a == b do
    do_squeeze([b | rest])
  end

  defp do_squeeze([a | rest]), do: [a | do_squeeze(rest)]
  defp do_squeeze([]), do: []

  def classify(text) when is_binary(text) do
    sig =
      text
      |> String.graphemes()
      |> Enum.chunk_every(2, 1, :discard)
      |> Enum.map(fn [a, b] -> relation(a, b) end)
      |> Enum.join()
      |> squeeze()

    case sig do
      "<" -> "A"
      ">" -> "D"
      "<>" -> "P"
      "><" -> "V"
      _ -> "X"
    end
  end
end

# Example Usage
IO.puts(Classifier.classify("abc"))    # A
IO.puts(Classifier.classify("cba"))    # D
IO.puts(Classifier.classify("abcba"))  # P
IO.puts(Classifier.classify("cbabc"))  # V
IO.puts(Classifier.classify("abdca"))  # X

File: WINGS_AI_Bootcamp/Python/ClassifyStrings.py
def relation(a: str, b: str) -> str:
    if a < b:
        return "<"
    elif a == b:
        return "="
    else:
        return ">"

def squeeze(s: str) -> str:
    if len(s) < 2:
        return s
    if s[0] == s[1]:
        return squeeze(s[1:])
    else:
        return s[0] + squeeze(s[1:])

def classify(text: str) -> str:
    sig = squeeze(''.join(relation(x, y) for x, y in zip(text, text[1:])))
    if sig == "<":
        return "A"   # Ascending
    elif sig == ">":
        return "D"   # Descending
    elif sig == "<>":
        return "P"   # Peak
    elif sig == "><":
        return "V"   # Valley
    else:
        return "X"   # None of the above

# Example Usage:
for s in ["abc", "cba", "abcba", "cbabc", "abdca"]:
    print(s, "->", classify(s))
